### PR TITLE
Add a new LSTM based classification algorithm

### PR DESCRIPTION
**Pull request by EvandroCT**
This  PR corresponds to the final code submission of my Google Summer of Code 2018 project. This doesn't mean, although, that I will no longer contribute to the COAL project.

It was introduced the code that allows one to use a trained Neural Network model to classify, pixel-by-pixel, an AVIRIS imagery. I've added a new script in the examples session to illustrate the use of a pre-trained binary classifier I created to  identify Schwertmannite pixels in AVIRIS-NG images. The model was trained using [this software](https://github.com/EvandroCT/dl-generalize) which adapts a RNN to take advantage of the sequencial nature of the bands to perform inferences with confidence. 

A more detailed documentation is coming soon.

Cheers!

### CODE REVIEW

File: examples/example_mineral_lstm.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+__all__ = []
+
+DEBUG = 1
+TESTRUN = 0
+PROFILE = 0
+
+
+input_filename = 'avng.jpl.nasa.gov/AVNG_2015_data_distribution/L2/ang20150420t182050_rfl_v1e/ang20150420t182050_corr_v1e_img.hdr'
+model_filename='../pycoal/tests/nn_models/lstm_Schwertmannite.h5'
```
Comment (lewismc) [inline examples/example_mineral_lstm.py:65]: This model need to be included in the pull request. Please include it!

File: pycoal/mineral.py [hunk 2]
```diff
@@ -142,10 +153,100 @@ def SAM(library, image_file_name, classified_file_name, scores_file_name=None, t
                 'map info': image.metadata.get('map info')
             })
 
+def avngDNN(image_file_name, classified_file_name, model_file_name, class_names=None, scores_file_name=None, in_memory=False):
```
Comment (lewismc) [inline pycoal/mineral.py:156]: This function name is pretty specific... has the code been tested against any AVIRIS-C imagery or only AVIRIS-NG?

File: pycoal/mineral.py [hunk 3]
```diff
@@ -142,10 +153,100 @@ def SAM(library, image_file_name, classified_file_name, scores_file_name=None, t
...
+    an image that holds all the SAM scores yielded for each pixel of the 
+    classified image. No score image is create if not provided.
+
+    In order to improve performance on systems with sufficient memory,
+    enable the optional parameter to load entire images.
+
+    Args:
+        image_file_name (str):          filename of the image to be classified
+        classified_file_name (str):     filename of the classified image
+        model_file_name (str):          filename of the Keras model used to classify
```
Comment (lewismc) [inline pycoal/mineral.py:171]: I thought we were going to use [Apache MXNet](https://mxnet.apache.org)? Regardless, we need to provide documentation on ***how*** a model is initially created.

File: pycoal/tests/mineral_test.py [hunk 4]
```diff
@@ -48,10 +48,7 @@ def _test_classify_image_teardown():
 def test_classify_image():
```
Comment (lewismc) [inline pycoal/tests/mineral_test.py:48]: It would also appear that you have not yet provided a Unit test for the NN classification option. Please add a unit test.

**Review by lewismc — CHANGES_REQUESTED** [review]
Hi @EvandroCT this is looking pretty good. Please see my comments. Right now I am unable to test the PR due to the absence of a trained model.

File: pycoal/mineral.py (hunk 2 above)
Comment (EvandroCT) [reply pycoal/mineral.py:156]: That's right. The name of the function is really misleading, because one can input a model that works with any shape of data, either AVIRIS-C, AVIRIS-NG or even a resampling approach. I'm going to change it.

File: pycoal/mineral.py (hunk 3 above)
Comment (EvandroCT) [reply pycoal/mineral.py:171]: Due to delays related to my unsuccessful trials to get other example scenes, there was not enough time to make also the MXNet models, but I'm going to make another callback function to work with them, too.

File: pycoal/tests/mineral_test.py (hunk 4 above)
Comment (EvandroCT) [reply pycoal/tests/mineral_test.py:48]: Sure!

File: examples/example_mineral_lstm.py (hunk 1 above)
Comment (EvandroCT) [reply examples/example_mineral_lstm.py:65]: @lewismc, My apologies, I forgot to include the model's folder into the PR. I'm going to include it right away!

**Comment by lewismc** [comment]
OK @EvandroCT let me know when this is ready to review again. Thank you

File: pycoal/mineral.py (hunk 3 above)
Comment (EvandroCT) [reply pycoal/mineral.py:171]: @lewismc , I've finished up testing the jupyter notebook tutorial cited in [my last week report](https://github.com/ESIPFed/GSoC/wiki/ISSUE-10-Reporting#week-13-6th-august--14th-august):

[https://www.lapig.iesa.ufg.br/drive/index.php/s/JbnPcl0GDHezPO0](https://www.lapig.iesa.ufg.br/drive/index.php/s/JbnPcl0GDHezPO0)

 This is part of the final documentation but I don't know where to place it within Pycoal's tree. Any suggestion?

File: pycoal/mineral.py (hunk 2 above)
Comment (lewismc) [reply pycoal/mineral.py:156]: Please go ahead and change the function name to something more descriptive.

File: pycoal/mineral.py (hunk 3 above)
Comment (lewismc) [reply pycoal/mineral.py:171]: Please just create a new repository and name it pycoal_examples, you can add it there. Thanks

Comment (EvandroCT) [reply pycoal/mineral.py:171]: Here's the repo, @lewismc : https://github.com/capstone-coal/pycoal_examples/blob/master/tutorial.ipynb

Comment (lewismc) [reply pycoal/mineral.py:171]: Great.
